### PR TITLE
#311

### PR DESCRIPTION
**Pull request by cmercer**
AbstractJacksonFieldSnipplet.java
- Updated to deal with superclass generic return type

JacksonResponseFieldSnippetTest.java
- Updated test objects to add super classes and interfaces
- Added genericSuperMethod() test

### CODE REVIEW

**Comment by loganm27** [comment]
Surprised that generics weren't supported out of the box...

**Review by loganm27 — APPROVED** [review]

File: spring-auto-restdocs-core/src/main/java/capital/scalable/restdocs/payload/AbstractJacksonFieldSnippet.java [hunk 1]
```diff
@@ -88,6 +91,20 @@ protected FieldDescriptors createFieldDescriptors(Operation operation,
...
         Type type = param.getGenericParameterType();
         if (type instanceof ParameterizedType) {
+            ParameterizedType parameterizedType = (ParameterizedType) type;
+            Type actualArgument = parameterizedType.getActualTypeArguments()[0];
+            if(actualArgument instanceof Class) {
+                return actualArgument;
+            } else if(actualArgument instanceof TypeVariable) {
+                TypeVariable typeVariable = (TypeVariable)actualArgument;
+                String variableName = typeVariable.getName();
+                Map<TypeVariable, Type> typeMap = GenericTypeResolver.getTypeVariableMap(param.getContainingClass());
```
Comment (scottresnik) [inline spring-auto-restdocs-core/src/main/java/capital/scalable/restdocs/payload/AbstractJacksonFieldSnippet.java:101]: Can't you just look up the type using the TypeVariable directly from the map?

**Comment by cmercer** [comment]
Generics were partially supported. 
Returning a typed List for example worked fine.
```
List<Item> getItems()
```
but getting the type from superclass/interfaces didn't work.

File: spring-auto-restdocs-core/src/main/java/capital/scalable/restdocs/payload/AbstractJacksonFieldSnippet.java (hunk 1 above)
Comment (cmercer) [reply spring-auto-restdocs-core/src/main/java/capital/scalable/restdocs/payload/AbstractJacksonFieldSnippet.java:101]: That is what was happening already.  
The MethodParameter param parameter is actually the Return "Parameter" (bad naming)
When you look it up on line 95 one of two things happen.  

Its either a Class, which means you have a "non-generic" return type like `List<Item>`.  This already worked fine.

Or its a TypeVariable, which means you have "generic" return type like `List<E>`.

The TypeVariable gives you a name and bounds, bounds being the case where you say `E extends Number`  Number becomes your lower bounds.  The main thing is you don't know what the actual "class" is.  For that you have to reference back to the "containingClass" and for that I used spring GenericTypeResolver to give me back the name -> class reference.

Comment (scottresnik) [reply spring-auto-restdocs-core/src/main/java/capital/scalable/restdocs/payload/AbstractJacksonFieldSnippet.java:101]: I see.  Makes sense.

**Review by scottresnik — APPROVED** [review]

**Pull request merged**